### PR TITLE
feat(kws): kws-streaming driver — pretrained Keyword Transformer models working in the browser

### LINKED ISSUE #72: task: kws-streaming driver module (Google Research kws_streaming, Traditional category)
## Context

`kws_streaming` (Google Research, `google-research/kws_streaming`, Apache-2.0) is
a Keras library of **streaming-aware KWS models** — dnn / cnn / gru / crnn /
ds_cnn / svdf / svdf_resnet / ds_tc_resnet (MatchboxNet) / bc_resnet — trained
non-streaming on Speech Commands and automatically converted to streaming
inference (internal or external state), then to TFLite. Published accuracies on
Speech Commands V2 (12 labels): bc_resnet_1 96.4% @ ~10K params, bc_resnet_2
97.6% @ 30K, ds_tc_resnet 98.0% @ 75K.

It belongs to the **Traditional Fixed-Class KWS** category (ADR-024 §2.1:
train + inference, new keywords require retraining, quantized TFLite/ONNX
export). Under the ADR-024 §1 decoupling rule and ADR-030, integrating it is a
**new driver module + panel**, with no edits to `kws-engine`.

Why it is a good fit for WakeStudio:

- Its `STREAM_EXTERNAL_STATE_INFERENCE` mode produces a **stateless** graph:
  all streaming buffers are graph inputs/outputs. That maps 1:1 onto
  `KWSBackend.processFrame(samples)` — feed one packet + the carried states,
  read logits + updated states back. No custom runtime needed.
- It is a **tiny-model** family (10K–75K params), which is exactly the MCU /
  app-class export target matrix (ADR-021), and it is Apache-2.0 **with no
  pretrained weights shipped** — so anything trained from it is commercially
  clean and passes the Phase-4 license gate (unlike the openWakeWord CC BY-NC-SA
  demo models).
- It closes the ADR-024 §2.1 gap: Traditional **training** currently has no
  upstream integration (`docs/kws-categories.md` §6 P0 lists ML-KWS-for-MCU /
  TorchKWS as intent only).

Upstream: https://github.com/google-research/google-research/blob/master/kws_streaming/README.md
Paper: "Streaming keyword spotting on mobile devices" (arXiv:2005.06720).

## Acceptance criteria

- [ ] `docs/modules/kws-streaming.md` written docs-first and human-reviewed
      (module spec per `docs/module-template.md`).
- [ ] New driver module `packages/modules/kws/streaming/`
      (`@wake-studio/module-kws-streaming`, `meta.id: kws-streaming`) with
      `spec/module.spec.json` validating against the contracts schema.
- [ ] `KWSStreamingBackend` implements `KWSBackend`, self-registers via
      `registerKwsBackend` with `category: 'traditional'` — **zero** edits to
      `packages/modules/kws/engine/core/` (guarded by the existing
      `decoupling.test.ts`).
- [ ] External-state streaming inference over onnxruntime-web (no new runtime
      dependency): states carried across `processFrame` calls, reset on
      `reset()`.
- [ ] Model + state shapes come from a **sidecar manifest** (`model.json`) so
      the driver is architecture-agnostic (one driver serves dnn…bc_resnet).
- [ ] `spec.train` follows ADR-031: pinned upstream
      `kws_streaming/train/model_train_eval.py`, **not forked**, plus a
      `standardize-results` adapter mapping the upstream run dir (`flags.json`,
      `labels.txt`, `tflite_stream_state_external/`) into the standard bundle.
- [ ] L1 tests: spec integrity + state-carry/reset logic (no model needed).
- [ ] Panel renders from the spec via module-kit (dual-layer, ADR-024 §4);
      registered in the web app + worker assembly seam.
- [ ] `LICENSES.md` row for `kws_streaming` (Apache-2.0) + Speech Commands
      (CC-BY-4.0) dataset.
- [ ] `docs/kws-categories.md` §6 P0 Traditional entry updated to point at this
      module.

## Out of scope (follow-ups)

- Actually training a shipped model + registering weights in
  `model-registry.json` (needs the Phase-5 training runner).
- ONNX conversion pipeline for the TFLite artifacts (build workflow) — tracked
  separately once the training path runs.
- Device/MCU export of `bc_resnet` (Phase 4, ADR-021).

### PR DESCRIPTION
**Pull request by dev-bobsong**
## Goal

Make **pretrained weights actually work in the web console** for the
`kws_streaming` model family (issue #72).

Four pretrained models now load and score in the browser:

| Registry id | Model | ONNX | Upstream top-1 | Our re-validation |
|---|---|---|---|---|
| `kws-streaming-kwt1` | Keyword Transformer, 1 head | 3.7 MB | 97.61% | **100.0%** (120/120) |
| `kws-streaming-kwt2` | Keyword Transformer, wider | 11.0 MB | 98.24% | **100.0%** (120/120) |
| `kws-streaming-kwt3` | Keyword Transformer, largest | 22.9 MB | 98.65% | **99.2%** (119/120) |
| `kws-streaming-att-mh-rnn` | CNN+biLSTM+multi-head attention | 6.0 MB | 98.48% | **99.2%** (119/120) |

Source: `ARM-software/keyword-transformer` (itself built on
`google-research/kws_streaming`), converted TFLite→ONNX by our own CI build.

## The design correction worth reviewing

My first pass implemented only upstream's **external-state streaming** path.
Then the actual published checkpoints turned out to be **non-streamable**
topologies (attention over the whole sequence) shipping **only**
`tflite_non_stream/` — so none of them could ever have loaded. Upstream cannot
convert those to streaming, so the window has to slide on our side.

The manifest therefore gained a `mode` discriminator:

- `streaming-external-state` — packet in → posterior + carried states out
  (unchanged; still the path for `dnn`…`bc_resnet`).
- `sliding-window` — the non-streaming graph over a 1 s window, re-evaluated
  every 100 ms. What ARM's weights need.

Validation is mode-aware: streaming mode *rejects* non-streamable topologies and
names the fix; sliding-window *accepts* them.

## How correctness is established

Each layer catches something the previous one cannot — this mattered, because
"the graph converted and ran" proves very little:

1. **Build-time, real audio** (`validate-kws-streaming.py`): runs each export
   over real Speech Commands clips, asserting argmax accuracy against the
   checkpoint's own `labels.txt` order, plus a silence check. A graph can convert
   cleanly and pass a zeros-input smoke test while being numerically wrong (bad
   weight layout, dropped transform, permuted outputs). Build fails below
   `min_accuracy`.
2. **L2, real artifact in Node**: manifest validates *and* matches the graph's
   real tensor names; 12 finite logits summing to 1 after softmax; silence fires
   nothing. Plus the alignment guard: **100 × 160-sample AFE frames pushed
   through `SlidingWindow` reproduce whole-clip inference logit-for-logit**. An
   off-by-one or a dropped sample would still yield plausible scores.
3. **L3, Playwright**: the real UI reaches `ready`, which happens only after the
   manifest cross-checked against the graph.

## Plumbing bug found and fixed

The KWS worker decided whether to load a backend by testing for the
**openwakeword model triple**, so any driver with a different URL shape was
silently skipped — engine `ready`, but deaf. Replaced with a registry capability
(`hasRequiredUrls`) that drivers declare, so the next backend needs no worker
edit (ADR-024 decoupling). The regression test for it is also what caught a
missing engine export.

Model + manifest travel as one registry entry (`url` + `manifestUrl`), so a graph
can never be paired with another model's manifest.

## Caveats recorded, not papered over

- **`sha256` is `null` on purpose.** `tf2onnx` output is not byte-reproducible:
  the same checkpoint produced 3 713 927 bytes locally and 3 713 945 in CI.
  Nothing verifies the field at runtime today; pinning hashes requires making
  the build deterministic first. Recording one now would just break on rebuild.
- **ARM weights are amber, not green, in `LICENSES.md`.** The repo's
  `LICENSE.txt` is Apache-2.0 and the README licenses "the source files"; the
  checkpoints are *data*, and no separate weights license is declared. So
  Apache-2.0 coverage is implied by inclusion, not stated. No NC clause exists
  anywhere and the corpus is CC-BY-4.0, so risk is low — but I flagged it for the
  Phase-4 commercial-export gate rather than claiming clean standing.

## Also

- Fixed a real gap in the generic build workflow: it declared only the `module`
  selector, so a module's `build.inputs` could never be overridden at dispatch
  (Actions has no dynamic `workflow_dispatch` inputs). Added an `inputs_json`
  escape hatch expanded into the `INPUT_<ID>` vars `build-module.mjs` already
  reads — benefits every module, documented in `docs/build-artifacts.md`.
- Added a build-artifact SOP checklist item: "the artifact is validated, not just
  built".
- Q-KS-3 answered (convert to ONNX — `tf2onnx` handled all four with no op
  coverage gaps, so we keep one inference stack and skip tfjs).

## Verification

- `pnpm -r test:unit` green (kws-streaming: 68 passed, 1 skipped)
- `pnpm -r typecheck` / `pnpm -r lint` clean (0 errors)
- All 27 Playwright e2e pass, incl. the new `kws-streaming.spec.ts`
- CI dispatched on this branch: **success** (run 31385944460)
- Build runs: 31382681352, 31382813639, 31384334242, 31385112437 (all success)

## Still open for you

- **Q-KS-1** (partly): default is `kwt1` (smallest). Should we also train a
  `bc_resnet_2` for the MCU tier? That is the only way to get a *streaming*
  model, since none is published.
- **Q-KS-2**: `preprocess: mfcc/micro` exports (feature extractor outside the
  graph) still rejected — matters for MCU parity.
- **Q-KS-4**: training panel ownership (this module vs the `training` module).

Closes #72

### CODE REVIEW

**Comment by dev-bobsong** [comment]
## Three reported bugs — three different causes

### 1. Keywords never trigger → `capturedAtMs` was in **seconds**

The real cause is not in this module at all. `AFEOutputFrame.capturedAtMs` was
assigned raw `AudioContext.currentTime` (seconds — the AFE's own latency math
multiplied it by 1000), while `TriggerDetector` compares that timestamp against
`minDurationMs` (300) and `cooldownMs` (2000). Triggering therefore required
**300 seconds** of continuous above-threshold score.

**This affected every backend** — openwakeword and plix included. Fixed at the
single source (worklet emits `currentTime * 1000`), AFE latency math corrected to
match, field documented as milliseconds.

> **My first hypothesis was wrong, and the test caught it.** I assumed pushing 0
> when `processFrame()` returns null collapsed the smoothed score between hops.
> It does not — the worker `return`s early on null, so the trigger never advances
> on those frames. `score-cadence.test.ts` keeps that refutation as an explicit
> test so the wrong theory can't come back.
>
> I still changed null-handling to a sample-and-hold, but for an honest reason: a
> hard 0 made the score *curve* sawtooth between 0 and the real score, which
> reads as a broken detector and hides real confidence.

### 2. Blank wake-word entries → an `options` contract mismatch

The ModuleSpec JSON schema declares `params[].options` as `string[]`, but
`ModuleParam.options` was typed `{value,label}[]`. The renderer read `.label` off
a string → `undefined` → blank entries.

**Also platform-wide**: afe-graph `topology`, kws-engine `executionProvider` and
training `target`/`backend` all use the string form. Fixed in the shared renderer
(`normalizeSelectOptions` in module-kit), with the contract type *and* JSON
schema updated to declare both shapes — the type was lying before.

Verified in-browser: wake word now reads `yes`; `wasm` / `WebGPU (faster)` render.

### 3. Only KWT-1 loads → the artifact only contained KWT-1

Not a code bug: my local assets came from an early single-checkpoint build, so
kwt2/kwt3/att_mh_rnn genuinely 404 while the registry advertised all four. I have
since rebuilt with all four (run 31455765355, success).

The code fix is the UX: a 404 now names the likely cause and gives the exact
commands. Verified live:

```
kws-streaming: manifest 'kwt3.json' is not present (404). The ONNX models are not
committed (ADR-011): run `node scripts/fetch-artifact.mjs kws-streaming`. If other
models load but this one 404s, the artifact was built for a subset of checkpoints -
rebuild with all of them: gh workflow run build.yaml -f module=kws-streaming ...
```

⚠️ **Note for you:** I could not finish downloading the 43 MB four-model artifact
here — the transfer stalls repeatedly on this connection. So I have verified
kwt1 end-to-end locally, and kwt2/kwt3/att_mh_rnn only in CI (converted +
validated at 98-100% on real audio). **Please run
`node scripts/fetch-artifact.mjs kws-streaming` and confirm all four load** — that
is the one claim I cannot close myself.

### Also

`sizeBytes` is now nominal (rounded): kwt1.onnx measured 3 713 927 / 3 713 945 /
3 713 930 bytes across three builds, so an exact count is wrong after any
rebuild. Doc sizes corrected to MiB to match what the UI renders.

### Verification

- `pnpm -r test:unit` green; kws-streaming 70 passed, +16 new tests overall
- `pnpm -r typecheck` / `lint` clean; **28/28** e2e pass
- CI on branch: success (31457398504)

**Comment by dev-bobsong** [comment]
## Inference threw on every frame → WASM EP pinned

Your console log was the whole answer. Two clues:

1. `ort-wasm-simd-threaded.**jsep**.mjs` — that's the **WebGPU** build.
2. The `VerifyEachNodeIsAssignedToAnEp` warning right before it — classic
   WebGPU graph partitioning.

**The bug:** onnxruntime-web's WebGPU (jsep) EP mis-executes the graph's
CLS-token `Slice` → `Squeeze`. The node squeezes axis 1 of `{1,1,64}` — which is
valid (axes initializer is int64 `[1]`, opset 17) — but jsep ignores the `axes`
input and tries axis 2, hence "must be 1 instead of 64".

**Why the driver was on WebGPU at all — my mistake.** I put
`executionProvider: "wasm"` in the *driver* spec, but `load()` receives the
**engine's** global `executionProvider`, which defaults to `webgpu`. The driver's
own preference was never consulted. So my "WASM by default" claim in the docs was
simply not true in practice.

**Fix:** the driver pins WASM regardless of what's requested — the same thing the
plix embedder already does for its own jsep op bug. No performance cost: these
graphs are 10K–750K params on a 1 s window every 100 ms, so WebGPU dispatch
overhead would dominate anyway. Verified the identical graph produces 12 finite
logits on WASM.

**Bonus lie fixed:** the worker hard-coded `gpuBackendLoaded = true` after any
successful load, so the UI showed "EP: WEBGPU" even when a backend had silently
used WASM. Backends now report `effectiveExecutionProvider` and the worker uses
that.

## The testing gap — this is the part worth your attention

**My L3 test was near-worthless and I should have caught that.** It asserted the
engine reached `ready`, which only proves *a session was created*. It passed
while **every single `run()` threw**.

And L2 couldn't save me either: onnxruntime-**node** has no jsep EP, so the same
graph ran perfectly there. A load-only browser test plus a Node inference test
leaves exactly this hole — the browser-only EP.

So I replaced L3 with a real inference spec that:
- drives the **actual driver** in a real browser and asserts a score comes out;
- asserts WASM is used when WebGPU is requested.

It needs a bundled page (`page.evaluate` can't resolve bare module specifiers), so
there's a harness under `apps/web/e2e-fixtures/` built **only** with
`E2E_HARNESS=1` — test scaffolding shouldn't ship to users.

**I checked the test actually catches the bug:** reverting the WASM pin makes it
fail with `Expected "wasm", Received "webgpu"`; restoring it passes.

Also: the spec's `executionProvider` now offers `wasm` **only** — offering
`webgpu` advertised a setting the driver overrides.

## Verification

- **30/30** e2e (2 new inference tests), all unit suites green, typecheck + lint clean
- CI on branch: success (31463873869)

## Please re-test

Triggering should now work. If it still doesn't fire, the next thing I'd want is
the **score curve** behaviour — whether the score rises when you speak (model
working, threshold/duration tuning needed) or stays flat (still an inference or
audio-routing problem). That distinction tells us which half to dig into.
